runtime/testlib_test.go

```diff
@@ -114,3 +138,697 @@ func TestExecuteScript(t *testing.T) {
 	require.Error(t, err)
 	assert.ErrorAs(t, err, &interpreter.TestFrameworkNotProvidedError{})
 }
+
+func TestInterpretMatcher(t *testing.T) {
+	t.Parallel()
+
+	t.Run("custom matcher", func(t *testing.T) {
+		t.Parallel()
+
+		script := `
+            import Test
+
+            pub fun main() {
+
+                let matcher = Test.NewMatcher(fun (_ value: AnyStruct): Bool {
+                     if !value.getType().isSubtype(of: Type<Int>()) {
+                        return false
+                    }
+
+                    return (value as! Int) > 5
+                })
+
+                assert(matcher.test(8))
+            }
+        `
+
+		storage := newTestLedger(nil, nil)
+
+		runtimeInterface := &testRuntimeInterface{
+			storage: storage,
+		}
+
+		_, err := executeScript(script, runtimeInterface)
+		require.NoError(t, err)
+	})
+
+	t.Run("custom matcher primitive type", func(t *testing.T) {
+		t.Parallel()
+
+		script := `
+            import Test
+
+            pub fun main() {
+
+                let matcher = Test.NewMatcher(fun (_ value: Int): Bool {
+                     return value == 7
+                })
+
+                assert(matcher.test(7))
```

Comment:
Just to clarify my understanding: `matcher` has non-generic type `AnyStruct{Test.Matcher}` here, correct?

If that's the case, then `matcher.test` could be called with a non-`Int` value here – is the run-time type checked somewhere? (I can't quite see it)

Reply:
Aha, good catch! This is not handled at the moment. The test-case I had added for this is flawed (it checks equality, so it fails anyway). Need to add a runtime check.
Enforcing that statically would have been much better, but I don't see a way to do that. Do you have any ideas?

Reply:
added the runtime check with 4792dd12f0565d7972d43f0fef7a50661afe69fb

Reply:
Nice! Thank you for also adding a test